src/Integration/WPCOMProxy.php

```diff
@@ -0,0 +1,215 @@
+		add_filter(
+			'woocommerce_rest_prepare_' . $object_type . '_object',
+			[ $this, 'check_item_is_syncable' ],
+			9,
+			3
+		);
+
+		add_filter(
+			'woocommerce_rest_prepare_' . $object_type . '_object',
+			[ $this, 'filter_metadata' ],
+			10,
+			3
+		);
+
+		add_filter(
+			'woocommerce_rest_' . $object_type . '_object_query',
+			[ $this, 'filter_by_metaquery' ],
+			10,
+			2
+		);
+	}
+
+	/**
+	 * Whether the data should be filtered.
+	 *
+	 * @param WP_REST_Request $request The request object.
+	 *
+	 * @return bool
+	 */
+	protected function should_filter_data( WP_REST_Request $request ): bool {
+		// WPCOM proxy will set the gla_syncable to 1 if the request is coming from the proxy and it is the Google App.
+		return $request->get_param( 'gla_syncable' ) === '1';
+	}
+
+	/**
+	 * Check if a single item is syncable.
+	 *
+	 * @param WP_REST_Response $response The response object.
+	 * @param mixed            $item     The item.
+	 * @param WP_REST_Request  $request  The request object.
+	 *
+	 * @return WP_REST_Response The response object updated.
+	 */
+	public function check_item_is_syncable( $response, $item, WP_REST_Request $request ): WP_REST_Response {
+		if ( ! $this->should_filter_data( $request ) ) {
+			return $response;
+		}
+
+		$route               = $request->get_route();
+		$pattern             = '/(?P<resource>[\w]+)\/(?P<id>[\d]+$)/';
+		$protected_resources = [
+			'products',
+			'coupons',
+		];
+		preg_match( $pattern, $route, $matches );
+
+		if ( ! isset( $matches['id'] ) || ! isset( $matches['resource'] ) || ! in_array( $matches['resource'], $protected_resources, true ) ) {
+			return $response;
+		}
+
+		$meta_data = $response->get_data()['meta_data'] ?? [];
+
+		foreach ( $meta_data as $meta ) {
+			if ( $meta->key === '_wc_gla_visibility' && $meta->value === 'sync-and-show' ) {
+				return $response;
+			}
+		}
+
+		return new WP_REST_Response(
+			[
+				'code'    => 'gla_rest_item_no_syncable',
+				'message' => 'Item not syncable',
+				'data'    => [
+					'status' => '403',
+				],
+			],
+			403
+		);
+	}
+
+	/**
+	 * Query items with specific args for example where _wc_gla_visibility is set to sync-and-show.
+	 *
+	 * @param array           $args    The query args.
+	 * @param WP_REST_Request $request The request object.
+	 *
+	 * @return array The query args updated.
+	 * */
+	public function filter_by_metaquery( array $args, WP_REST_Request $request ): array {
+		if ( ! $this->should_filter_data( $request ) ) {
+			return $args;
+		}
+
+		$post_type         = $args['post_type'];
+		$post_type_filters = self::$post_types_to_filter[ $post_type ];
+
+		if ( ! isset( $post_type_filters['meta_query'] ) || ! is_array( $post_type_filters['meta_query'] ) ) {
+			return $args;
+		}
+
+		$args['meta_query'] = [ ...$args['meta_query'] ?? [], ...$post_type_filters['meta_query'] ];
+
+		return $args;
+	}
+
+	/**
+	 * Filter the metadata of an object.
```

Comment:
I think it would be more accurate something like:

"Filter the response metadata returning only those prefixed with _wc_gla"